### PR TITLE
Nef_3: performance in binop_intersection_tests

### PR DESCRIPTION
**Pull request by GilesBathgate**
## Summary of Changes

Within `Nef_3/include/CGAL/Nef_3/binop_intersection_tests.h`, the current code calculates bounding boxes for intersecting elements multiple times when performing intersection tests. While calculating the bounding boxes is typically a very fast operation, it does impact performance, particularly when the number of elements is large. The current code implements two `std::vector<Nef_box>` called `a` and `b` which are repeatedly cleared and recomputed.

At first, it was considered it might be intentional to re-calculate because the SNC structure is updated via the callbacks. However, in the binary operation, the source operands are const, and the result is a new value. The code uses a `const_cast` to cast away constness from the parameters.  The first step was to make the necessary changes to ensure that the parameters and callbacks are const, thus removing the undesirable `const_cast`'s

The current pair of nef box vectors could then be replaced with four nef box vectors that persist for the function scope and are reused. Reserving the size of the vectors improves allocation speed.

Some redundant code and headers were removed. The code relating to CGAL_NEF3_BOX_INTERSECTION_CUTOFF was removed since it cannot compile because of the typo made in https://github.com/CGAL/cgal/commit/97769c439e16db569089f0a5147916a6c3680e65 where a trailing comma was added on line 220:

https://github.com/CGAL/cgal/blob/97769c439e16db569089f0a5147916a6c3680e65/Packages/Nef_3/include/CGAL/Nef_3/binop_intersection_tests.h#L218-L220

I will provide some benchmark results in upcoming comments.

## Release Management

* Affected package(s): Nef_3
* Issue(s) solved (if any): performance
* Feature/Small Feature (if any):
* License and copyright ownership: Returned to CGAL authors.

### CODE REVIEW

**Comment by afabri** [comment]
I have read the commits one by one and they make sense.     Is there a measurable performance gain?

File: Nef_3/include/CGAL/Nef_3/Binary_operation.h [hunk 1]
```diff
@@ -167,14 +171,14 @@ class Binary_operation : public CGAL::SNC_decorator<Map> {
             typename Selection,
             typename Association>
   class Intersection_call_back :
-    public SNC_point_locator::Intersection_call_back
+    public CGAL::SNC_point_locator<SNC_decorator>::Intersection_call_back
```
[the remark this answers is not in the document]
Comment (GilesBathgate) [reply Nef_3/include/CGAL/Nef_3/Binary_operation.h:174]: The reason for the change is that it needs a `CGAL::SNC_point_locator<SNC_const_decorator>` here, and not `SNC_point_locator` which is CGAL::SNC_point_locator<SNC_decorator>. I am not sure what typedef you mean for me to add.

[the remark this answers is not in the document]
Comment (afabri) [reply Nef_3/include/CGAL/Nef_3/Binary_operation.h:174]: I am confused. You inverted needs and not, right?    I had in mind `SNC_point_locator`  and `SNC_const_point_locator`.

[the remark this answers is not in the document]
Comment (GilesBathgate) [reply Nef_3/include/CGAL/Nef_3/Binary_operation.h:174]: `SNC_const_point_locator` does make sense. I will add a typedef.

**Comment by GilesBathgate** [comment]
> Is there a measurable performance gain?

These are my results I will need to double check these but it looks like 1.8% improvment:

version | time (ms) | version| time (ms) |  
-- | -- | -- | -- | --
baseline: | 35268 | contender: | 33675 |  
baseline: | 36239 | contender: | 36039 |  
baseline: | 35897 | contender: | 35348 |  
baseline: | 33253 | contender: | 35776 |  
baseline: | 35525 | contender: | 32732 |  
baseline: | 35946 | contender: | 34708 |  
baseline: | 35683 | contender: | 34827 |  
baseline: | 33142 | contender: | 32777 |  
  | 35119.125 |   | 34485.25 | 1.8%

**Comment by GilesBathgate** [comment]
Results again, this time with CGAL_NO_ASSERTIONS, the performance is only ~1.2%

version | time (ms) | version | time (ms) |  
-- | -- | -- | -- | --
baseline: | 13049 | contender: | 12588 |  
baseline: | 13425 | contender: | 13409 |  
baseline: | 13176 | contender: | 13339 |  
baseline: | 13275 | contender: | 12786 |  
baseline: | 12833 | contender: | 12821 |  
baseline: | 13507 | contender: | 13665 |  
baseline: | 13660 | contender: | 13609 |  
baseline: | 13578 | contender: | 13577 |  
baseline: | 13603 | contender: | 12802 |  
baseline: | 13738 | contender: | 13665 |  
  | 13384.4 |   | 13226.1 | 1.18%

**Review by afabri — APPROVED** [review]

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.6-Ic-169

**Pull request merged**